sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,833 @@
+        # received is pre-binned.
+        # predicting is faster on pre-binned data, so we want early stopping
+        # predictions to be made on pre-binned data. Unfortunately the scorer_
+        # can only call predict() or predict_proba(), not raw_predict(), and
+        # there's no way to tell the scorer that it needs to predict binned
+        # data.
+        self._in_fit = True
+
+        # bin the data
+        if self.verbose:
+            print("Binning {:.3f} GB of data: ".format(X.nbytes / 1e9), end="",
+                  flush=True)
+        tic = time()
+        self.bin_mapper_ = _BinMapper(max_bins=self.max_bins, random_state=rng)
+        X_binned = self.bin_mapper_.fit_transform(X)
+        toc = time()
+        if self.verbose:
+            duration = toc - tic
+            print("{:.3f} s".format(duration))
+
+        self.loss_ = self._get_loss()
+
+        self.do_early_stopping_ = (self.n_iter_no_change is not None and
+                                   self.n_iter_no_change > 0)
+
+        # create validation data if needed
+        self._use_validation_data = self.validation_fraction is not None
+        if self.do_early_stopping_ and self._use_validation_data:
+            # stratify for classification
+            stratify = y if hasattr(self.loss_, 'predict_proba') else None
+
+            X_binned_train, X_binned_val, y_train, y_val = train_test_split(
+                X_binned, y, test_size=self.validation_fraction,
+                stratify=stratify, random_state=rng)
+            if X_binned_train.size == 0 or X_binned_val.size == 0:
+                raise ValueError(
+                    'Not enough data (n_samples={}) to '
+                    'perform early stopping with validation_fraction='
+                    '{}. Use more training data or '
+                    'adjust validation_fraction.'.format(
+                        X_binned.shape[0],
+                        self.validation_fraction)
+                )
+            # Predicting is faster of C-contiguous arrays, training is faster
+            # on Fortran arrays.
+            X_binned_val = np.ascontiguousarray(X_binned_val)
+            X_binned_train = np.asfortranarray(X_binned_train)
+        else:
+            X_binned_train, y_train = X_binned, y
+            X_binned_val, y_val = None, None
+
+        if self.verbose:
+            print("Fitting gradient boosted rounds:")
+
+        # initialize raw_predictions: those are the accumulated values
+        # predicted by the trees for the training data. raw_predictions has
+        # shape (n_trees_per_iteration, n_samples) where
+        # n_trees_per_iterations is n_classes in multiclass classification,
+        # else 1.
+        n_samples = X_binned_train.shape[0]
+        self._baseline_prediction = self.loss_.get_baseline_prediction(
+            y_train, self.n_trees_per_iteration_
+        )
+        raw_predictions = np.zeros(
```

Comment:
It might be easier to understand using `np.repeat` to initialize?

Reply:
Not sure if this what you mean but this doesn't work

```py
raw_predictions = np.repeat(self._baseline_prediction, (self.n_trees_per_iteration_, n_samples))
```

```
operands could not be broadcast together with shape (1,) (2,)
```

Reply:
Whoops, I did not catch that `self._baseline_prediction` was not of `n_samples` size already. Then, it will not work.